### ISSUE TITLE
Using .latex method causes edit handler to fire twice

### ISSUE THREAD

**Issue by YakovL** [mathquill/mathquill#601]
So, here's a very basic self-explanatory (I believe) example showing the bug. Type `window.mathField.latex("smth")` in console, that will fire the edit handler which writes the message twice (hence it is called twice). I also tried to create a [jsFiddle](https://jsfiddle.net/rgw7z413/), but it requires loading libs through https which seems to be unavailable (or I should use some other address?).

```
<html>
  <head>
    <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.7.2/jquery.min.js"></script>
    <!-- works with libs loaded localy, too -->
    <script src="http://mathquill.com/lib/mathquill.js"></script>
    <link rel="stylesheet" href="http://mathquill.com/lib/mathquill.css" type="text/css" />
  </head>
  <body>
	<!-- if window.mathField.latex("smth") is called, edit handler fires twice, see the console -->
	<span id="mathquill"></span>
	<script>
	  var mathFieldSpan = document.getElementById('mathquill');
	  var MQ = MathQuill.getInterface(2);
	  window.mathField = MQ.MathField(mathFieldSpan, {
        spaceBehavesLikeTab: true,
		handlers: {
		  edit: function() { // useful event handlers
		    console.log("edit handler is called");
		  }
		}
	  });
	</script>
  </body>
</html>
```

This is a bug, right? If so, a quick inline fix would be very appreciated (for my current project, I'll probably change the local lib code or use another workaround until this is fixed in the main lib).

**Comment by laughinghan** [comment]
It's clearly suboptimal, but I wouldn't consider it a bug, it's documented as being called when the field _might_ have changed. That is, there's no guarantee the field has changed between calls to `edit`, but it is guaranteed that `edit` is called whenever the field does change. The docs used to be more explicit about this but I thought "might" was clear enough, perhaps they should be changed back? https://github.com/mathquill/mathquill/commit/9dc72e11f2df1aa3e2da49bef1dd44732ad302ff#diff-04c6e90faac2675aa89e2176d2eec7d8L212

This isn't the only case where `edit` may be called even though the field hasn't change. Another example is if you put the cursor at the very beginning (left edge) of the math field and backspace: nothing happens, but the `edit` handler is called anyway. It'd be nontrivial to inexpensively, architecturally guarantee that `edit` isn't called unless the math field has actually been edited (but such a patch would be welcome, of course), so we make the API consumer explicitly cache the LaTeX and on `edit`, check against that (if they care).

(An argument could be made the other way, I suppose: MathQuill should hide the "expensive" check and then in the future, obviate it and transparently speed up everyone; by making the API consumer do it, even when MathQuill obviates it, the API consumer won't automatically get the performance bump.)

**Comment by The-Penultimate-Defenestrator** [comment]
I think MathQuill should expose a built-in implementation of your
recommendation to the API consumer, even if it’s slower. Both methods could
be kept.

**Comment by YakovL** [comment]
Ok, let's consider some pragmatics before deciding anything about what's the right thing to do with this issue.

My case is the following (may be you'll suggest a better way to handle it). I've created a editor of formulae which has two fields: MathQuill field and (plain) LaTeX field (there is, in fact, the third field containing CodeCogs image, which is synced on `change` of the LaTeX field). More precisely, I have forked an editor which contained only LaTeX field. The MathQuill and LaTeX fields are supposed to get synced.

In order to avoid loops of syncing, I introduced the mathSyncedOnce param, and used it like this:

```
	var latexEditor = document.getElementById('mathjaxSrc');
	// this is used to prevent loops between LaTeX and MathQuill change handlers:
	var mathSyncedOnce = false;

	// apply MathQuill to the main field, update LaTeX field on edit
	var MQ = MathQuill.getInterface(2);
	var mathQuillField  = document.getElementById('editable-math'),
	    mathQuillEditor = MQ.MathField(mathQuillField, {
		spaceBehavesLikeTab: true, //# ??
		handlers: {
			edit: function() {
				if(!mathSyncedOnce) {
					mathSyncedOnce = true;
					latexEditor.value = mathQuillEditor.latex(); // update LaTeX field
					$(latexEditor).change(); // fire updating of CodeCogs
				}
				else mathSyncedOnce = false;
			}
		}
	});

	//...

	$(latexEditor).on("keyup change",function(){
		if(!mathSyncedOnce) {
			mathSyncedOnce = true;
			mathQuillEditor.latex(latexEditor.value); //// in fact, this causes firing edit handler twice!
		}
		else mathSyncedOnce = false;
	});
```

But then I was suggested to avoid the behaviour when in LaTeX editor one starts typing "a + b" and once (s)he types the space symbol after "a" it gets removed right after it is typed (so the user may try this several times without success). Tracking the origin of the problem revealed the MathQuill behaviour we're discussing here.

As a workaround, I changed the usage of the param (and renamed it) in the fashion that can be seen below:

```
	var latexEditor = document.getElementById('mathjaxSrc');
	// this is used to prevent loops between LaTeX and MathQuill change handlers:
	var mathNotSyncedOnce = true, MQ_DONE_ALREADY = 2;
	// MQ_DONE_ALREADY trick is used to "fix" MathQuill behaviour:
	// see https://github.com/mathquill/mathquill/issues/601

	// apply MathQuill to the main field, update LaTeX field on edit
	var MQ = MathQuill.getInterface(2);
	var mathQuillField  = document.getElementById('editable-math'),
		mathQuillEditor = MQ.MathField(mathQuillField, {
		spaceBehavesLikeTab: true, //# ??
		handlers: {
			edit: function() {
				if(mathNotSyncedOnce == true) { // not MQ_DONE_ALREADY

					mathNotSyncedOnce = false;
					latexEditor.value = mathQuillEditor.latex(); // update LaTeX field
					$(latexEditor).change(); // fire updating of CodeCogs
				}
				else if(mathNotSyncedOnce == MQ_DONE_ALREADY)
					mathNotSyncedOnce = true;
				else mathNotSyncedOnce = MQ_DONE_ALREADY;
			}
		}
	});

	// ...

	$(latexEditor).on("keyup change",function(){
		if(mathNotSyncedOnce) { // either true or MQ_DONE_ALREADY

			mathNotSyncedOnce = false;
			mathQuillEditor.latex(latexEditor.value);
		}
		else mathNotSyncedOnce = true;
	});
```

This works fine when `.latex` causes firing MathQuill `edit` handler _exactly_ twice (which is for now perfectly reproducible in my tests), but this is not lib-behaviour-change-proof (if it fires once, then the next change won't be synced; if it fires thrice, the "eat space" behaviour won't be prevented).

What would you recommend to avoid this issue?

If the `edit` handler got event object as a param, I could check its type (sync only if it's some "keypress" or "keyup" or "keydown"; do the same in the latex field handler) and avoid these weird solutions, but as far as I can see from docs, `edit` handler doesn't get any info about the event.

**Comment by laughinghan** [comment]
Would your problems be solved if you kept around the last known LaTeX of the math field and compared against that in the `edit` handler? This would be effectively equivalent to suppressing duplicate/spurious `edit` calls right? Except you could go further and update your variable just before calling `.latex()` on the math field.

The plan for a while has in fact been to transition the event handlers to take in an event object, with a `.from` key (or `.provenance` but my friends think that's way too pretentious) whose values could be `event` or `api`, similar to the `source` parameter to [QuillJS event handlers](http://quilljs.com/docs/events/) (no relation in spite of the similar name).

**Comment by YakovL** [comment]
It seems that this trick helps only to prevent _double_-syncing of _one_ field, but doesn't suppress (single) _self_-syncing (of the LaTeX field through the MathQuill field which eats spaces). At least this doesn't work properly:

    	var mathQuillField  = document.getElementById('editable-math'),
    		mathQuillEditor = MQ.MathField(mathQuillField, {
    		spaceBehavesLikeTab: true,
    		handlers: {
    			edit: function() {
    				if(mathNotSyncedOnce == true && mathQuillEditor.latex() != oldMQvalue) {
    
    					oldMQvalue = mathQuillEditor.latex();
    					mathNotSyncedOnce = false;
    					latexEditor.value = mathQuillEditor.latex(); // update LaTeX field
    					$(latexEditor).change(); // fire updating of CodeCogs
    				}
    				else mathNotSyncedOnce = true;
    			}
    		}
    	});
    ...
    	$(latexEditor).on("keyup change",function(){
    		if(mathNotSyncedOnce && latexEditor.value != oldLatexValue) {
    			oldLatexValue = latexEditor.value;
    
    			mathNotSyncedOnce = false;
    			mathQuillEditor.latex(latexEditor.value);
    		}
    		else mathNotSyncedOnce = true;
    	});
    	var oldLatexValue = latexEditor.value;

(it doesn't eat space immediately but rather once the next symbol is typed). But that's no surprise: once the symbol after the space is typed in LaTeX, it updates MathQuill and since there actually are some changes, it modifies MathQuill back. Actually, I don't see how this can help. I can try to create a simple .html with all necessary stuff inside to reproduce the issue so that you can test it if you like..

**Comment by laughinghan** [comment]
Oh I see, I was hoping you could just set `oldLatexValue = latexEditor.value` before updating the math field, but you're right, MathQuill will normalize the LaTeX, so it won't necessarily be the same. It does happen synchronously though, so something like this should work, right? 
```js
    var currentlyCallingAPI = false;
    var mathQuillSpan  = document.getElementById('editable-math');
    var mathQuillEditor = MQ.MathField(mathQuillSpan, {
        spaceBehavesLikeTab: true, // BTW see below
        handlers: {
            edit: function() {
                if (currentlyCallingAPI) return;

                latexEditor.value = mathQuillEditor.latex(); // update LaTeX field
                $(latexEditor).change(); // fire updating of CodeCogs (?)
            }
        }
    });
...
    $(latexEditor).on('keydown keypress paste', function() {
        setTimeout(function() {
            currentlyCallingAPI = true;
            mathQuillEditor.latex(latexEditor.value);
            currentlyCallingAPI = false;
        });
    });
```

That does kinda suck though, I'm hoping to solve this to be something like:
```js
    var mathQuillSpan  = document.getElementById('editable-math');
    var mathQuillEditor = MQ.MathField(mathQuillSpan, {
        spaceBehavesLikeTab: true, // BTW see below
        handlers: {
            edit: function(e) {
                if (e.from === 'API') return;

                latexEditor.value = mathQuillEditor.latex(); // update LaTeX field
                $(latexEditor).change(); // fire updating of CodeCogs (?)
            }
        }
    });
...
    $(latexEditor).on('keydown keypress paste', function() {
        setTimeout(function() {
            mathQuillEditor.latex(latexEditor.value);
        });
    });
```

By the way, `spaceBehavesLikeTab` is a [documented config option](http://mathquill.readthedocs.org/en/latest/Config/#spacesbehavesliketab), you can omit it if it's not the behavior you want.

**Comment by LennyLip** [comment]
@laughinghan How to save cursor position after mathQuillEditor.latex(latexEditor.value) is calling?
